src/measure_theory/borel_hierarchy.lean

```diff
@@ -0,0 +1,440 @@
+`s : set (set α)` is the generating family.
+
+The main difference with the classical definition is that this hierarchy starts
+at level 0: Π⁰₀ coincides with `s`
+augmented with `∅` and `univ` (intended to be a countable basis of a topology) and
+Σ⁰₀ is the empty family.
+-/
+inductive sigma0_pi0_rec {α : Type u} (s : set (set α)) :
+  ordinal.{u} → bool → set α → Prop
+| basic (x ∈ s) : sigma0_pi0_rec 0 tt x
+| empty         : sigma0_pi0_rec 0 tt ∅
+| univ          : sigma0_pi0_rec 0 tt univ
+| compl {i x}   : sigma0_pi0_rec i ff x → sigma0_pi0_rec i tt xᶜ
+| union {i} (f : ℕ → set α) (g : ℕ → ordinal.{u}) :
+    (∀ n, g n < i) → (∀ n, sigma0_pi0_rec (g n) tt (f n)) → sigma0_pi0_rec i ff (⋃ n, f n)
+
+/--
+The family of (boldface) Σ⁰ᵢ pointsets, which are countable unions of Π⁰ⱼ sets
+(given by the function `pointclasses.pi0` below) of smaller index. The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+-/
+def sigma0 : set (set α) := sigma0_pi0_rec s i ff
+
+/--
+The family of (boldface) Π⁰ᵢ pointsets, which are the complements of Σ⁰ᵢ sets
+(given by the function `pointclasses.sigma0` above). The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+
+When the ordinal argument is `0`, `pi0` returns the generating family `s`.
+-/
+def pi0 : set (set α) := sigma0_pi0_rec s i tt
+
+lemma sigma0_pi0_rec_def' {b : bool} :
+  sigma0_pi0_rec s i b = if b then pi0 s i else sigma0 s i :=
+by { unfold pi0 sigma0, cases b; refl }
+
+/-- Base case for `sigma0`. -/
+@[simp] lemma sigma0_zero : sigma0 s 0 = ∅ :=
+begin
+  unfold sigma0,
+  ext x, rw [mem_empty_iff_false, iff_false],
+  intro hx,
+  induction' hx with _ _ _ _ _ _ _ f g glt hf IH,
+  exact ordinal.not_lt_zero (g 0) (glt 0)
+ end
+
+/-- `sigma0` comprises unions of previous `pi0`. -/
+lemma sigma0_eq_Union_pi0:
+  sigma0 s i = set.range (λ (f : ℕ → ⋃ j (hij : j < i), pi0 s j), ⋃ n, (f n).1) :=
+begin
+  rcases classical.em (i=0) with rfl | hi; unfold sigma0, rw sigma0_pi0_rec_def',
+  { apply eq.symm, simp [range_eq_empty, ordinal.not_lt_zero] },
+  { ext x, split; intro hx,
+    { induction' hx with _ _ _ _ _ _ i f g glt hf,
+      existsi λn, (⟨f n, _⟩ : ↥⋃ j < i, pi0 s j),
+      { simp only [eq_self_iff_true] },
+      { rw mem_Union,
+      use g n,
+      rw mem_Union,
+      exact ⟨glt n, hf n⟩ } },
+    { cases hx with f hf,
+      rw ← hf,
+      dsimp only at hf,
+      choose g hg using λn, (mem_Union.mp (f n).property),
+      simp only [mem_Union,exists_prop] at hg,
+      apply sigma0_pi0_rec.union _ g,
+      exact λ n, (hg n).1,
+      unfold pi0 at hg,
+      exact λ n, (hg n).2 } }
+end
+
+lemma pi0_subset_sigma0 (hik : i < k) :
+  pi0 s i ⊆ sigma0 s k :=
+begin
+  simp only [sigma0_eq_Union_pi0,hik],
+  intros x hx,
+  apply mem_range.mpr,
+  have hxU : x ∈ ⋃ j < k, pi0 s j,
+  { simp only [mem_Union,exists_prop],
+    use i,
+    exact ⟨hik,hx⟩ },
```

Comment:
When you have everything at hand, you can give the full proof in one line: `exact ⟨i, hik,hx⟩`.

Reply:
I tried to do this at every place, but there's one spot where I decided to avoid this, which is the subproof that starts [here](https://github.com/leanprover-community/mathlib/compare/borel_hierarchy#diff-09ea7dca4ea866411209acae79260ea987186ddf20b678ed694056a7afa0e80aR341). Otherwise I would have a line too long.